src/backend/renderer/element/surface.rs

```diff
@@ -305,21 +307,29 @@ where
 }
 
 /// A single surface render element
-pub struct WaylandSurfaceRenderElement<R> {
+pub struct WaylandSurfaceRenderElement<R: Renderer> {
     id: Id,
     location: Point<f64, Physical>,
     alpha: f32,
-    surface: wl_surface::WlSurface,
     renderer_type: PhantomData<R>,
     kind: Kind,
+
+    view: Option<SurfaceView>,
+    current_commit: CommitCounter,
+    buffer: Option<Buffer>,
+    buffer_scale: i32,
+    buffer_transform: Transform,
+    buffer_dimensions: Option<Size<i32, BufferCoords>>,
+    damage: DamageBag<i32, BufferCoords>,
```

Comment:
Instead of `DamageBag` this could use `DamageSnapshot` which provides a read-only view of the damage in the `DamageBag` and was introduced to support exactly such use-cases. The snapshot can be obtained from `DamageBag::snapshot()`.